meyfa/ts-node-test
# feat: Support custom extensions via env variable

# Summary

* Add SUPPORTED_EXTENSIONS environment variable
* Split FS operations out into their own module
* Add tests


# Custom Extensions

Adding the ability to supply a custom `SUPPORTED_EXTENSIONS` environment variable gives the user more power when setting up their test runner. For example, a user may choose to write their test files adjacent to their source files and give it the suffix `.test.ts`. This change will allow a user to supply `SUPPORTED_EXTENSIONS=.test.ts` to selectively run TypeScript files within their source directory, rather than maintaining a separate test directory structure.

This is implemented within the `getSupportedExtensions` function in the `index.ts` file.


# FS Module

Filesystem operations have been shifted out of `index.ts` and into `fs.ts`.
The `resolveTestPaths` function has been updated to allow the supply of an array of allowed extensions through a new `extensions` parameter, rather than directly accessing the hard-coded array of allowed extensions. This also makes the function more testable.

The `resolveTestPaths` and `walkPath` functions have been updated to include a `readDirectory` function. In regular use, this defaults to the already used `readdir` function, however by including this parameter, it allows for other functions or mechanisms to be used to read a directory. This is primarily done to make the code more testable, but could also be used in the future to perform further pruning of the directory structure in advance to remove or reduce wasted `readDirectory` calls on files or paths that will never be of interest.


# Tests

* Ensure that the `getSupportedExtensions` function returns the expected built-in supported extensions
* Ensure that the `getSupportedExtensions` function returns the list of extensions supplied by the `SUPPORTED_EXTENSIONS` environment variable if it is present
* Ensure that the `walkPath` function successfully walks a basic directory structure with files and folders
* Ensure that the `resolveTestPaths` function successfully walks a basic directory structure and retrieves only the files of interest
*
